apps/emqx_durable_storage/src/emqx_ds_cache.erl

```diff
@@ -0,0 +1,510 @@
+                seqno => Seqno,
+                last_seen_key => LastSeenKey
+            }),
+            do_fetch_cache(Next, DB, Stream, LastSeenKey, Seqno + 1, BatchSize, _Acc = []);
+        _ ->
+            %% race condition: previous key doesn't match; GC may have run.
+            false
+    end.
+
+-spec do_fetch_cache(
+    '$end_of_table' | ?EOS_KEY(emqx_ds:stream()) | ?KEY(emqx_ds:stream(), emqx_ds:message_key()),
+    emqx_ds:db(),
+    emqx_ds:stream(),
+    emqx_ds:message_key(),
+    seqno(),
+    _Remaining :: pos_integer(),
+    [{emqx_ds:message_key(), emqx_types:message()}]
+) ->
+    cache_fetch_result().
+do_fetch_cache(
+    '$end_of_table' = Key, _DB, Stream, OrigLastSeenKey, _ExpectedSeqno, _Remaining, Acc
+) ->
+    with_last_key(Acc, Stream, Key, OrigLastSeenKey);
+do_fetch_cache(
+    ?EOS_KEY(Stream) = Key, _DB, Stream, OrigLastSeenKey, _ExpectedSeqno, _Remaining, Acc
+) ->
+    with_last_key(Acc, Stream, Key, OrigLastSeenKey);
+do_fetch_cache(Key, _DB, Stream, OrigLastSeenKey, _ExpectedSeqno, Remaining, Acc) when
+    Remaining =< 0
+->
+    with_last_key(Acc, Stream, Key, OrigLastSeenKey);
+do_fetch_cache(Key, DB, Stream, OrigLastSeenKey, ExpectedSeqno, Remaining, Acc) ->
+    ?tp(ds_cache_lookup_enter, #{key => Key, stream => Stream, seqno => ExpectedSeqno}),
+    case ets:lookup(?cache(DB), Key) of
+        [#entry{key = ?KEY(Stream, DSKey), seqno = ExpectedSeqno, message = Message}] ->
+            NextKey = ets:next(?cache(DB), Key),
+            NewAcc = [{DSKey, Message} | Acc],
+            do_fetch_cache(
+                NextKey, DB, Stream, OrigLastSeenKey, ExpectedSeqno + 1, Remaining - 1, NewAcc
+            );
+        _ ->
+            %% Either:
+            %%   i)   There's no entry;
+            %%   ii)  The streams don't match;
+            %%   iii) The seqnos don't match.
+            with_last_key(Acc, Stream, Key, OrigLastSeenKey)
+    end.
+
+-spec with_last_key(
+    [{emqx_ds:message_key(), emqx_types:message()}],
+    emqx_ds:stream(),
+    '$end_of_table' | ?EOS_KEY(emqx_ds:stream()) | ?KEY(emqx_ds:stream(), emqx_ds:message_key()),
+    emqx_ds:message_key()
+) ->
+    cache_fetch_result().
+with_last_key(Acc, Stream, Key, OrigLastSeenKey) ->
+    case Acc of
+        [{LastKey, _} | _] ->
+            Batch = lists:reverse(Acc),
+            ?tp(ds_cache_hit, #{stream => Stream, batch => Batch, last_key => LastKey}),
+            {ok, LastKey, Batch};
+        [] when Key =:= ?EOS_KEY(Stream) ->
+            ?tp(ds_cache_eos_found, #{}),
+            {ok, end_of_stream};
+        [] ->
+            %% Possible if entry was removed just before we started iterating, or if only
+            %% the last seen key of the iterator is contained in the cache.  If the latter
+            %% case, then we should return a cache "hit" with empty results; otherwise,
+            %% clients that are up to date with their iterators will start spamming RPCs
+            %% again.
```

Comment:
TODO: add some tests to check this.

Reply:
Should rethink and revert this: if a topic filter is later removed, any client whose last seen key is in the cache might get forever stuck...

Reply:
One idea is to return `false` if the stream is not currently being tracked, otherwise return an empty batch to prevent the client from doing the RPC.